src/Control/Effect/Sum.hs

```diff
@@ -22,59 +18,35 @@ instance (HFunctor f, HFunctor g) => HFunctor (f :+: g)
 instance (Effect f, Effect g)     => Effect   (f :+: g)
 
 
-type Member (sub :: (* -> *) -> (* -> *)) sup = MemberAt (PathTo sub sup) sub sup
-
-inj :: forall sub m a sup . Member sub sup => sub m a -> sup m a
-inj = inj' @(PathTo sub sup)
-
-prj :: forall sub m a sup . Member sub sup => sup m a -> Maybe (sub m a)
-prj = prj' @(PathTo sub sup)
-
-
--- | Construct a request for an effect to be interpreted by some handler later on.
-send :: (Member effect sig, Algebra sig m) => effect m a -> m a
-send = alg . inj
-{-# INLINE send #-}
-
-
-type family FromJust (maybe :: Maybe a) :: a where
-  FromJust ('Just a) = a
-
-type family (left :: Maybe k) <|> (right :: Maybe k) :: Maybe k where
-  'Just a <|> _       = 'Just a
-  _       <|> 'Just b = 'Just b
-  _       <|> _       = 'Nothing
-
-type family Prepend (s :: j -> k) (ss :: Maybe j) :: Maybe k where
-  Prepend s ('Just ss) = 'Just (s ss)
-  Prepend _ 'Nothing   = 'Nothing
-
-data L a
-data R a
-
-type family PathTo' (side :: * -> *) (sub :: (* -> *) -> (* -> *)) sup :: Maybe * where
-  PathTo' s t t         = 'Just (s ())
-  PathTo' s t (l :+: r) = Prepend s (PathTo' L t l <|> PathTo' R t r)
-  PathTo' _ _ _         = 'Nothing
-
-type family PathTo sub sup where
-  PathTo t t         = ()
-  PathTo t (l :+: r) = FromJust (PathTo' L t l <|> PathTo' R t r)
-
-class MemberAt path (sub :: (* -> *) -> (* -> *)) sup | path sup -> sub where
-  inj' :: sub m a -> sup m a
-  prj' :: sup m a -> Maybe (sub m a)
-
-instance MemberAt () t t where
-  inj' = id
-  prj' = Just
-
-instance MemberAt path t l => MemberAt (L path) t (l :+: r) where
-  inj' = L . inj' @path
-  prj' (L l) = prj'  @path l
-  prj' _     = Nothing
-
-instance MemberAt path t r => MemberAt (R path) t (l :+: r) where
-  inj' = R . inj' @path
-  prj' (R r) = prj'  @path r
-  prj' _     = Nothing
+class Member (sub :: (* -> *) -> (* -> *)) sup where
+  inj :: sub m a -> sup m a
+  prj :: sup m a -> Maybe (sub m a)
+
+instance Member sub sub where
+  inj = id
+  prj = Just
+
+instance {-# OVERLAPPABLE #-}
+         Member sub (l1 :+: l2 :+: r)
+      => Member sub ((l1 :+: l2) :+: r) where
+  inj = reassoc . inj where
+    reassoc (L l)     = L (L l)
+    reassoc (R (L l)) = L (R l)
+    reassoc (R (R r)) = R r
+  prj = prj . reassoc where
+    reassoc (L (L l)) = L l
+    reassoc (L (R l)) = R (L l)
+    reassoc (R r)     = R (R r)
```

Comment:
I’m a bit concerned about the runtime cost of this, but it makes for dramatically better error messages. However, since `HasIn` now exists, that might be a moot point, since our error messages usually don’t mention `Member` anyway. (Only the one for statically known signatures with more than one element does, and it’s pretty easy to build a better error for that case using `GHC.TypeLits`’ `TypeError`.)

I am at any rate intending to split `Member` up into separate classes for injection and projection since the only thing needing projection is interposition. So I’ll take a look at the compile- and run-time performance of this machinery at the same time.